Hrt 52 create account

lucycy98 opened the pull request:
### Purpose

<!---
Describe the problem or feature.
Consider providing an overview of why the work is taking place (with any relevant links); don’t assume familiarity with the history.
-->

<!---
Link to the related issues [KEY-000](link)
e.g [HRT-27](https://softeng761team5.atlassian.net/browse/HRT-27)
--->

Created sign up page for users. Fills in user details such as email, username, password, full name, institution, role, department.
Calls create user endpoint upon clicking create account button.
sets errors: 
- duplicate username
- confirm password doesn't match password
- not all fields are filled

[HRT-52](https://softeng761team5.atlassian.net/browse/HRT-52)

<!-- []() -->
<!-- []() -->

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!--
If this PR contains a breaking change, describe the impact and migration path for existing applications below.
-->

### Screenshots before and after this change (required for UI changes)

#### Does this PR involve a UI change?

- [x] Yes
- [ ] No

#### Screenshot(s) of UI after this PR

screenshot (zoomed out to display full panel) 
<!---
Required if there is UI change
--->
![image](https://user-images.githubusercontent.com/34847224/94551003-264d5000-02b1-11eb-9db4-ef620d8a8069.png)

![image](https://user-images.githubusercontent.com/34847224/94551043-37965c80-02b1-11eb-894c-097e83546c25.png)

lucycy98 committed:
feat(signup): added frontend for create account

lucycy98 committed:
feat(users): updated user endpoint to send error when username already exists

lucycy98 committed:
feat(signup): calling create user endpoint with checks e.g duplicate username

gargi-bansal commented:
@lucycy98 looking really good. I'll test it properly with the backend tomorrow to check the errors and everything pop up. Two small requests: Could the create account button height be the same as the login button height, and the page is missing the "Already a member. Log in!" button from the designs so would be nice to have that in. 
<img width="720" alt="Screen Shot 2020-09-30 at 12 18 47 AM" src="https://user-images.githubusercontent.com/42624632/94551872-8395d100-02b2-11eb-809b-3c3c1f1328ff.png">

lucycy98 committed:
fix(signup): added links to login, fixed styling for consistency

lucycy98 commented:
> @lucycy98 looking really good. I'll test it properly with the backend tomorrow to check the errors and everything pop up. Two small requests: Could the create account button height be the same as the login button height, and the page is missing the "Already a member. Log in!" button from the designs so would be nice to have that in.
> <img alt="Screen Shot 2020-09-30 at 12 18 47 AM" width="720" src="https://user-images.githubusercontent.com/42624632/94551872-8395d100-02b2-11eb-809b-3c3c1f1328ff.png">

@gargi-bansal thanks, i've just added that :)

gargi-bansal commented on src/styles/Signup.css:
```diff
@@ -0,0 +1,79 @@
+  box-shadow: 0px 0px 0px 0px;
+  font-size: 24px;
+  margin-top: 30px;
+  border-radius: 10px;
```
If you're adding this styling here, then the Login.css login button also needs a border-radius of 10px

gargi-bansal commented on src/styles/Signup.css:
```diff
@@ -0,0 +1,79 @@
+}
+
+.Account-Button {
+  height: 55px;
```
Should be 60px like in Login

gargi-bansal commented on src/styles/Signup.css:
```diff
@@ -0,0 +1,79 @@
+  background: white;
+  box-shadow: 4px 4px 10px #212121;
+  width: 600px;
+  height: 1360px;
```
Probably good to add another 60px to the height in Signup and Login for the panels, so the button height is not minimised.

lucycy98 replied:
will make it consistent

gargi-bansal commented on src/pages/Signup.tsx:
```diff
@@ -0,0 +1,239 @@
+
+  const onCreateClick = async (): Promise<void> => {
+    if (
+      !username &&
```
these should all be OR instead of AND, as the error does not pop up correctly :)

gargi-bansal commented on src/pages/Signup.tsx:
```diff
@@ -0,0 +1,239 @@
+      setError(ALL_FIELDS_SET);
+      return;
+    } else if (password !== confirmPassword) {
+      return;
```
Could there be an error string for this that you could set, like in the previous if-statement. It would nice to give the user as much feedback as possible about where they went wrong.

Nvm, it seems like its there, it seems like you're not using setError() like you were with the other errors. It should probably be consistent and use setError()

lucycy98 replied:
I wanted the password doesn't match error to be dynamic when the user is typing rather than when the user clicks on the button :)

gargi-bansal replied:
ohhh that makes a lot more sense, yea it looks better when it's dynamic. I just couldn't understand how the code worked

gargi-bansal reviewed (changes requested):
The form looks great and the post request and logging in works well too! I added a few requests in-line on the PR and I also have a few requests for the error handling side of things.

It seems that you're showing errors in two places which is okay, except the "username taken" error should probably appear at the same place as the "passwords don't match error" since it's only specific to the top half of the form. Or you could just show all the errors all the way at the bottom. It's up to you how you want to fix it as both approaches are fine.

Another thing I did on the login page was the inputs relevant to the error shown would highlight in orange when the Login button was clicked. If you have a look at the code and play around with the UI you'll see what I mean. It would be nice if this page could do the same.

Oh also, I've added useEffects in every page to automatically redirect the user to the Login page or Dashboard based on whether they already have a cookie. You should probably copy in the one from Login.

Sorry for the super long comment. Hopefully, it shouldn't take too long to fix!

lucycy98 committed:
fix(signup): add redirect to dashboard if cookies

lucycy98 committed:
fix(signup login): fixed consistency styling

lucycy98 committed:
fix(signup): error messages in same place, styling of input during errors

lucycy98 commented:
@gargi-bansal addressed all your issues :) 
- note that the password match error will show when the user is typing (before they even click on the create account button, hence why it is a direct variable comparison)

preetpatel commented on server/users/users.controller.js:
```diff
@@ -12,8 +12,12 @@ router.route("/").post(bodyParser.json(), async (request, response) => {
-  } catch (error) {
-    return response.status(400).send(error);
+  } catch (e) {
+    console.log(e);
```
somewhere in your next PR, try remove this console log.. ill still approve it this time tho 🥇

gargi-bansal committed:
feat(front end): fix styling of panel height and button height

fix styling of panel height and button height for Login and Sign Up pages to be consistent

gargi-bansal reviewed (approved):
lgtm
